sympy/algebras/quaternion.py

```diff
@@ -244,26 +271,47 @@ def to_euler(self, seq):
         if not symmetric:
             a, b, c, d = a - c, b + d, c + a, d - b
 
-        # calculate angles
-        half_sum = atan2(b, a)
-        half_diff = atan2(d, c)
-
-        angle_j = 2 * atan2(sqrt(c * c + d * d), sqrt(a * a + b * b))
-        # alternatively, we can use this to avoid the square root:
-        # angle_2 = acos(2*(a*a + b*b)/(a*a + b*b + c*c + d*d) - 1)
-
-        angle_i = half_sum + half_diff
-        angle_k = half_sum - half_diff
+        if avoid_square_root:
+            n2 = self.norm()**2 if symmetric else 2 * self.norm()**2
+            angles[1] = acos((a*a + b*b - c*c - d*d) / n2)
+        else:
+            angles[1] = 2 * atan2(sqrt(c * c + d * d), sqrt(a * a + b * b))
+
+        # Check for singularities in numerical cases
+        angle_test = angles[1]
+        case = 0
+        if angle_test.is_number:
+            if is_eq(angle_test, S.Zero):
+                case = 1
+            if is_eq(angle_test, S.Pi):
+                case = 2
+
+        if case == 0:
+            if angle_addition:
+                angles[0] = atan2(b, a) + atan2(d, c)
+                angles[2] = atan2(b, a) - atan2(d, c)
+            else:
+                angles[0] = atan2(b*c + a*d, a*c - b*d)
+                angles[2] = atan2(b*c - a*d, a*c + b*d)
+
+        else:  # any degenerate case
+            warnings.warn('Singularity case, setting third angle to zero')
```

Comment:
Please don't add warnings like this. It is better just to give an error or something.